### PR TITLE
ci: better naming for various steps/jobs

### PR DESCRIPTION
**Pull request by agilgur5**
### Motivation

- partly inspired by #11670
  - as well as seeing users be confused by some steps
    - `git diff --exit-code` had [some confusion](https://github.com/argoproj/argo-workflows/pull/10806#issuecomment-1537913872) in particular
  - and me just reading more closely and wanting better descriptions (as I am working on a few CI changes)
    - for example, I did not even realize that the steps after an E2E test failure were for debugging purposes until I read the CI code itself
      - GH runs some post action hooks etc, so thought that was just clean-up or something

### Modifications

- add names to more steps & jobs in GH Actions Workflows
  - specific names for all `git diff --exit-code` as per above
  - names for Docs steps
  - names PR Title check steps
  - names for Snyk dependency scans workflow
  - CI workflow names:
    - all E2E test failure debugging steps per above
    - add a "get" (as in `kubectl get`) in front of the ones that just log out a resource for clarity as well
    - Java and Python SDK set-up/install steps
    - a few other clarifications / improvements
    - additions where there were no names

- also add some new lines and comments in a few places for clarity & readability

- reordering of fields
  - follow a consistent order of `name` -> `if` -> `needs`
  - then `uses`, `steps`, or `run` depending on the context

### Verification

- Purely stylistic and DX changes, no semantic changes
- CI on this PR should pass with all the new names too 🙂

### CODE REVIEW

**Comment by agilgur5** [comment]
Ok top-level workflow names and job names don't look great when they're long. Gonna make some modifications...

**Comment by agilgur5** [comment]
Hmmm required checks evidently require the name to be exactly the same instead of the key 😕 

See screenshot:
Screenshot 2023-09-05 at 9 09 54 PM

I'm going to revert the "Unit Tests" and "argoexec-image" name changes then, especially as those did not add altogether much

**Comment by Yaminyam** [comment]
image
Because the actions list is divided by the yaml file name, you should be aware that two actions with the same name may result in confusion.
I think it would be a good idea to delete the existing list when there are enough logs of newly changed docs actions and the old logs are no longer needed.

**Comment by agilgur5** [comment]
> Because the actions list is divided by the yaml file name, you should be aware that two actions with the same name may result in confusion.

Well that's annoying and confusing 😕
To be fair though, I almost never use the [Actions tab](https://github.com/argoproj/argo-workflows/actions) of the repo and normally just look at checks on commits and PRs.

> I think it would be a good idea to delete the existing list when there are enough logs of newly changed docs actions and the old logs are no longer needed.

I don't know if this is actually possible 😐 [This action](https://github.com/argoproj/argo-workflows/actions/workflows/default-branch-check.yaml) hasn't been run in 10 months but still appears in the list.
If there is such a capability (idk), I don't have permissions to delete any of these unfortunately.

So not sure that we can actually do anything about this, it's a GitHub issue 😕

**Comment by Yaminyam** [comment]
That's right, this is a GitHub issue and discussions are continuing, but there is no response from GitHub.
https://github.com/orgs/community/discussions/26256

So the only way to delete the meaningless list from the workflow list is to delete all logs.
https://stackoverflow.com/questions/57927115/delete-a-workflow-from-github-actions/67000032#67000032

image
As shown in the photo above, github actions logs cannot be checked after 90 days anyway, so it is okay to delete all logs older than 90 days.
I am using the script above to delete and manage logs that are no longer used due to changes in the yaml file.

**Comment by Yaminyam** [comment]
I often look at the actions tab, and the problem when the meaningless list becomes large is that you have to go in to check which of the workflows with the same name is currently in use.
When the number of workflows increases, it may be difficult to find a valid workflow because the workflows are sorted alphabetically.

**Comment by agilgur5** [comment]
> So the only way to delete the meaningless list from the workflow list is to delete all logs.
> https://stackoverflow.com/questions/57927115/delete-a-workflow-from-github-actions/67000032#67000032

I doubt I have permissions to do this, but maybe an approver does.
These [2](https://github.com/argoproj/argo-workflows/actions/workflows/default-branch-check.yaml) [actions](https://github.com/argoproj/argo-workflows/actions/workflows/security.yaml) can be pruned.

`docs.yaml` can't be pruned as it's the new file name here and contains logs from everything other than this PR (future old logs)

**Comment by Yaminyam** [comment]
Even if a lot of logs are piled up, they cannot be accessed after 90 days or more, so I think it is okay to delete the existing logs because they become meaningless after 90 days after the 90-day pr is applied.
But ultimately, this is a matter for those in power to decide. :D

**Comment by agilgur5** [comment]
> Even if a lot of logs are piled up, they cannot be accessed after 90 days or more

If this PR is merged, I would say it would be ok to delete `gh-pages.yaml` after 90 days (although PRs that are based on an old branch will still have it). But I don't have permission to do this regardless

File: .github/workflows/changelog.yaml
Comment (terrytangyuan) [inline .github/workflows/changelog.yaml:14]: What's the purpose of this change?

File: .github/workflows/changelog.yaml
Comment (terrytangyuan) [inline .github/workflows/changelog.yaml:33]: Same here

Comment (terrytangyuan) [reply .github/workflows/changelog.yaml:33]: And other places in this PR

File: .github/workflows/ci-build.yaml
Comment (terrytangyuan) [inline .github/workflows/ci-build.yaml:115]: Isn't it already intuitive by using this action?

File: .github/workflows/ci-build.yaml
Comment (terrytangyuan) [inline .github/workflows/ci-build.yaml:171]: This is too implementation specific and may change

File: .github/workflows/snyk.yml
Comment (terrytangyuan) [inline .github/workflows/snyk.yml:30]: Not necessary. It's already under a `node` section

**Review by terrytangyuan — COMMENTED** [review]
Please avoid unnecessary changes

File: .github/workflows/ci-build.yaml
Comment (agilgur5) [reply .github/workflows/ci-build.yaml:115]: I did not know why Python was being installed, given that it's not a language used inside of Argo. The SDK has a test here, and I mention the SDK in the name

File: .github/workflows/changelog.yaml
Comment (agilgur5) [reply .github/workflows/changelog.yaml:33]: there's no change here, it's just a newline at EOF, which GitHub and many editors point out. Mine automatically adds it when missing in any file

File: .github/workflows/changelog.yaml
Comment (agilgur5) [reply .github/workflows/changelog.yaml:14]: See my opening comment, which explicitly says:

> * reordering of fields
>     * follow a consistent order of `name` -> `if` -> `needs`
then `uses`, `steps`, or `run` depending on the context

File: .github/workflows/ci-build.yaml
Comment (agilgur5) [reply .github/workflows/ci-build.yaml:171]: Not sure what you mean, it already said "MinIO/MySQL pods", which I had no clue why it was doing that before, so I added "Failure debug" to be explicit.
I also didn't know what "MinIO/MySQL pods" meant. Versus, I run an SRE team and if you say "`[kubectl] get` a Pod", everyone knows what means. If you say "MinIO/MySQL pods", I would say "what about them?".
I can change it to `describe` which would be even more explicit.

There's two lines of code here, so yes, I would very much expect that the name would need changing if the two lines change. I don't think that's a bad thing, that means the name is explicit about what the two lines do and there is no guesswork to be had when looking at the CI logs, which normally auto-collapse these under the name.

File: .github/workflows/snyk.yml
Comment (agilgur5) [reply .github/workflows/snyk.yml:30]: It makes more sense in context; the below screenshot just says "Snyk / node":

Screenshot_2023-09-26-10-36-32-04_40deb401b9ffe8e1df2f1cc5ba480b12

That is not entirely clear what it does since Snyk has many functionalities (see also automated PRs #11844, scanning Docker images, etc). This makes it very explicit at first glance

**Comment by agilgur5** [comment]
> Please avoid unnecessary changes
 
I would disagree that these are unnecessary. I had several iterations and trial-and-error testing on this, as can be seen in the commit log and first two comments.

It seemed like you mostly didn't like the reordering and may have misinterpreted those as removals?

Per my opening comment, that was a consistency change:

> * reordering of fields
>   * follow a consistent order of `name` -> `if` -> `needs`
>   * then `uses`, `steps`, or `run` depending on the context

would be great if we had an automated formatter to sort fields in the same order, but I'm not sure that exists, so did so manually for now.
It's quite a bit easier on the eyes when they are consistent as opposed to all over the place, which makes it harder to follow and/or find the fields that I'm looking for. `name` first acts similarly to a comment, then `if` and `needs` are conditionals that say whether the rest is relevant (in code, the rest would be nested under these) or can be skipped.

Same as [in-line comment](https://github.com/argoproj/argo-workflows/pull/11752#discussion_r1337275193)

**Comment by agilgur5** [comment]
ah poop looks like this already ended conflicting with #11782. That was one of the reasons I asked for reviews [over Slack](https://cloud-native.slack.com/archives/C0510EUH90V/p1694726319948949?thread_ts=1694644337.627309&cid=C0510EUH90V) as renames and reorderings can be really easy to conflict with

**Review by terrytangyuan — APPROVED** [review]

File: .github/workflows/ci-build.yaml
Comment (agilgur5) [reply .github/workflows/ci-build.yaml]: I updated it to `describe`

**Comment by agilgur5** [comment]
Rebased and fixed merge conflicts! fingers crossed those are the last ones before this get merged 🤞

**Comment by terrytangyuan** [comment]
Build failed

**Comment by agilgur5** [comment]
Yea [docs](https://github.com/argoproj/argo-workflows/actions/runs/6322548346/job/17168375993?pr=11752) just timed out on checkout. That's happened a few times recently in CI actually.
The other one is [a test failure](https://github.com/argoproj/argo-workflows/actions/runs/6322548348/job/17168505793?pr=11752) that I've actually never seen before (`TestJSONVariables`).

But my changes don't affect either of those, so those are flakes in context. I can do an empty commit but I was hoping you could retry those @terrytangyuan

**Comment by agilgur5** [comment]
one of the flakey hooks tests is [failing](https://github.com/argoproj/argo-workflows/actions/runs/6341809982/job/17226399531?pr=11752) now 😕

**Pull request merged**
